### DISCUSSION TITLE
Linux and Windows support

### DISCUSSION THREAD

**Discussion by cbenhagen** [firebase/flutterfire#5557] [Feature request]
Options are:
- [C++ SDK](https://github.com/firebase/firebase-cpp-sdk)
- Writing / Adopting existing dart-native implementations like [firedart](https://github.com/cachapa/firedart).

The second option might be harder but in the long run it might be easier to maintain.

Please show your support for this feature request by upvoting [this idea](https://firebase.uservoice.com/forums/948424-general/suggestions/46591750-desktop-windows-linux-macos-support-for-flutte). Thanks!

**Comment by stuartmorgan-g** [comment]
> * [C++ SDK](https://github.com/firebase/firebase-cpp-sdk)

Note the following, from [the documentation for that SDK](https://firebase.google.com/docs/cpp/setup#desktop-workflow):
> Caution: Firebase C++ SDK desktop support is a beta feature. This feature is intended only for workflows during the development of your game, not for publicly shipping code.

I.e., it's an SDK for testing mobile games on desktop, rather than a desktop SDK.

**Reply by efraespada** [reply]
If Firebase C++ SDK desktop becomes `stable`, would it be integrated as expected?

**Comment by cbenhagen** [comment]
I guess pure Dart implementations would be the way to go then.

**Reply by cedvdb** [reply]
I'd say something lower level that all current sdk's can use would be the way to go. Wasm comes to mind

**Comment by salihgueler** [comment]
If we decide(e.g. I am planning to work on this) to tackle this, should we create separate issues per product and work on them accordingly?

**Comment by cbenhagen** [comment]
@salihgueler I think one meta issue and discussion in the PR would work as well but It's not up to me to decide. How do you plan to implement it?

**Comment by nilsreichardt** [comment]
An advantage of creating multiple tickets is that we can track which firebase package is how often requested.

**Comment by shivamdev-lgtm** [comment]
Is there any work being done on desktop support (other than MacOS)?

**Comment by creativecreatorormaybenot** [comment]
It seems like discussions cannot be used for linking to pull requests. I suggest creating individual issues for the packages and platforms.

**Comment by jlubeck** [comment]
Is there any progress here?

**Comment by dgurudot** [comment]
Any updates for windows and linux support?

**Comment by cedvdb** [comment]
The dart native option sounds like it has disadvantages:
  - looks like a monumental task
  - it's one more library doing the same thing as others, except in another language
  - there is the c++ sdk, might as well work on that one which will also support other languages, not just dart

**Comment by tanzil114** [comment]
Can someone please give an update about the current status for Windows support. Not having windows support is causing a lot of issues for us, so need to know how long it will take for Windows support to be covered. Thanks!
cc: @Salakar

**Reply by shivamdev-lgtm** [reply]
I don't think it's coming anytime soon

**Reply by tanzil114** [reply]
Thanks. well, I hope it's added soon as not having Firebase support is a deal-breaker for windows on Flutter

**Comment by barbalex** [comment]
Our app has 33 dependencies. 3 do not have windows support yet. Two of them are not vital (geolocation). The only blocking dependency left that does not support windows is firebase.

**Comment by worldwidee** [comment]
almost we are in 2022 and there is no windows support. The Windows version of my app is waiting for 6 months because i built with firebase for all platforms. "firedart" extension is giving nothing to me

**Reply by IchordeDionysos** [reply]
You can build an abstraction for Firebase? And use the Firebase APIs :)

**Comment by cedvdb** [comment]
If firedart existed, it would allow for tree shaking of the firebase js, doesn't it ?

**Reply by ghost** [reply]
personnaly I used the REST API, firedart have some limitation.

**Comment by dgurudot** [comment]
It will do code forking and also any update to firebase requires pull and it is not a true stream which makes the app not responsive and real time

**Comment by worldwidee** [comment]
And update please?

**Comment by richard457** [comment]
Please give up an update so we know how to plan.

**Comment by cedvdb** [comment]
Would a solution involving C++ sdk and FFI allow tree shaking ?  In other words, do the libraries used by dart ffi get tree shaken ? It would suprise me if the answer is yes, but who knows

One could run the FFI-gen on the C++ sdk as a starting point

**Comment by ghost** [comment]
IMO, fastest way is to wrap the C++ code to get something spun up ASAP. When/If that's successful and there's package size issues, etc, the pure dart implementation can be done. The Windows/Linux Flutter market isn't huge yet, but this would allow it to start to grow with the intention of switching it over at a later point, if need be.

**Comment by pr-Mais** [comment]
Hello everyone, we're happy to share that we started working on a Dart implementation to support Windows and Linux, currently it's core and auth, still in preview but your feedback and contributions are very welcomed!

http://github.com/invertase/flutterfire_desktop

**Reply by cedvdb** [reply]
What is the underlying sdk used ? Is it the C one ?

**Reply by barbalex** [reply]
Seems they are using the rest api: https://invertase.io/blog/announcing-flutterfire-desktop

**Reply by pr-Mais** [reply]
Yes it's using the REST API

**Reply by cedvdb** [reply]
What does that mean in this case ? 

Firebase is more than a client server communication. Especially at the firestore level. I understand the development has not started yet for firestore but a REST API is a bit nebulous here, if you mean make HTTP requests with a predefined structure. 

I just feel like it's odd to mention that it's using the REST api. Like it could have used something else ?

That's fantastic news nonetheless

**Reply by ghost** [reply]
Firebase have a complexe model of data and if you use the REST API you should create a parser in dart to transform api data into object. It's juste an example but if firebase upgrade his rest API you should rework your parser :').

If an official lib do that for you with many extra things useful for Firebase, I'm happy.
Actually you need to handle auth by your side too which can take some time to work and could be achieve by a lib ;)

**Reply by worldwidee** [reply]
I am happy to see there are works for firebase desktop. Firebase is perfect db for create app which is working on cross platform. I hope you guys can develop other firebase features soon 💪💪♥️

**Reply by pr-Mais** [reply]
There's logic happening inside the packages so it's not just a wrapper for the REST API, all functionalities of firebase core & auth is being supported following the same platform interface of FlutterFire, so for example, you can listen to a stream of events about the user auth state, exactly like other platforms, completely handled by the package.

Having a cross-platform SDK for Dart is a long-term goal, and many challenges regarding that is yet to be figured out, but the focus currently is primarily on Linux and Windows.

For Firestore, just like other Firebase SDKs, using gRPC API would probably be the way to go, but it's still not yet planned and the focus now is on Core & Auth.

**Reply by cedvdb** [reply]
I personally don't think this is the right direction. The work has to be an effort from google to have a cross platform source. There is no way this is going to reach production stage for all firebase products, when those products are ever evolving with teams dedicated to all of those.

**Comment by shivamdev-lgtm** [comment]
Is there any **official** update about this?

**Comment by LauDijksterhuis** [comment]
I would love windows & linux support in Firebase. 
But even if it's not planned, can this page: https://firebase.flutter.dev/ be updated to show all platforms?

It's really useful to see which packages support which platform. 🙂

**Reply by pr-Mais** [reply]
The support for Linux and Windows is in-progress. Currently, we are working on Auth which is still in preview, and development is ongoing, the documentation on how to use the finished parts is continuously being added to README for each package e.g. [firebase_auth_desktop README](https://github.com/invertase/flutterfire_desktop/tree/main/packages/firebase_auth/firebase_auth_desktop#readme).

**Comment by mstfkhazaal** [comment]
Are there any updates?

**Comment by developerfix** [comment]
Any updates for windows platform?

**Comment by shigomany** [comment]
Is there any information when there will be updates on firebase for Windows?

**Comment by cbenhagen** [comment]
@shigomany you can follow this issue: https://github.com/invertase/flutterfire_desktop/issues/52

**Reply by shigomany** [reply]
Thanks

**Comment by cbenhagen** [comment]
@ everyone subscribed to this discussion: The Flutter and Firebase teams rely on the feedback they gather to shape their roadmaps. So to support Firebase and Linux and Windows please do the following:
- Create a feature request for Firebase C++ and/or Dart versions [here](https://firebase.google.com/support/troubleshooter/report/features).
- Hit the thumbs-up button on [this issue](https://github.com/firebase/firebase-cpp-sdk/issues/442). 
- Make sure to leave comments about the need for Firebase desktop support in the quarterly Flutter surveys.

Thanks!

**Comment by cbenhagen** [comment]
@marb2000 / @puf do you have any news on the plans Google has for officially supporting Firebase on other Desktop platforms than macOS? Already partial support for the most important plugins like Firestore / Realtime Database would help a lot.

**Comment by hwkim1127** [comment]
actually I found a pull request here.
check it out.

[feat(core): add support for Windows via Firebase C++ desktop SDK #10496](https://github.com/firebase/flutterfire/pull/10496#issue-1592915387)

**Comment by cbenhagen** [comment]
@everyone show your support for this feature request by upvoting [this idea](https://firebase.uservoice.com/forums/948424-general/suggestions/46591750-desktop-windows-linux-macos-support-for-flutte). Thanks!

**Comment by hwkim1127** [comment]
Well I think this should be asked here

https://github.com/firebase/firebase-cpp-sdk/issues

Here 
https://firebase.uservoice.com/forums/948424-general/suggestions/46591750-desktop-windows-linux-macos-support-for-flutte

And here
https://firebase.google.com/support/troubleshooter/report/features

**Comment by michaeljajou** [comment]
Would really benefit for FlutterFire on Windows - we need this! 🙌

**Reply by stMerlHin** [reply]
Same for linux.

**Comment by cbenhagen** [comment]
As of today Windows support using the C++ SDK has landed for [auth](https://github.com/firebase/flutterfire/pull/11089), [firestore](https://github.com/firebase/flutterfire/pull/11516) and [storage](https://github.com/firebase/flutterfire/pull/11617) plugins. Based on this effort it should be quite easy to also support Linux. I am looking forward to an official statement and hope it will soon added to the README as well.

**Reply by melWiss** [reply]
same goes here mate

**Comment by hwkim1127** [comment]
I really hope there will be a support for Realtime Database as well...
Or it has to be done separately with Rest API for Desktop...

**Reply by hwkim1127** [reply]
actually... I am now using SSE for RTDB when it comes to Windows app.
I like it better than the [firebase_database](https://pub.dev/packages/firebase_database), since I don't have to handle wierd parsing stuff with Object

**Comment by alexvoina** [comment]
is there a roadmap / estimate for Crashlytics on Windows? I've seen it on a roadmap for 2022 as "not started". How do we stand today on this?

**Reply by puf** [reply]
Crashlytics itself only provides native SDKs for iOS (and other Apple platforms) and Android at the moment. so there's no way for the plugins in this repo to add support for Windows by wrapping a native SDK. 

I recommend filing a feature request [with Firebase support](https://firebase.google.com/support/troubleshooter/report) and on [firebase.uservoice.com](https://firebase.uservoice.com) to add support for additional platforms.

**Comment by stMerlHin** [comment]
Does the actual C++ SDK support linux ? If so, why there is not linux support yet on flutterfire side if we ignore the fact that flutterfire team is busy ? After all windows is supported and it does use the C++ SDK...

**Reply by stMerlHin** [reply]
+1

**Comment by raymon-io** [comment]
Is there any update or roadmap for Windows and Linux?

**Comment by melWiss** [comment]
+1

**Comment by elvinasn** [comment]
Looking up for this... 😢

**Comment by soufianebenyaala** [comment]
Any updates for linux support?

**Comment by m-berto** [comment]
Asking for Linux support. Tnx

**Comment by soufianebenyaala** [comment]
Asking for Linux support. Tnx

**Comment by NonymousMorlock** [comment]
Almost 5 years, still waiting
+1 Bump. Asking for Linux support

**Comment by ndevev** [comment]
Asking for Windows support.
Thanks.

**Reply by NonymousMorlock** [reply]
I thought we already had windows support, from v 2.11.0 [changelog](https://pub.dev/packages/firebase_core/changelog#2110)

**Comment by melWiss** [comment]
We as a community would like to understand why the delay, not blaming or anything. And is there anything we can help with? Like if everyone is spawning random packages to add this support, those will never be completed.

If there's a clear one repo to this support, documents, anything, clear roadmap, I believe that the community will be able to help more efficiently especially if there's also a clear leadership on this, we really want to help each other here.

**Comment by o-glethorpe** [comment]
Are you guys planning a big party for the 5-year anniversary?

**Comment by hitchcliff** [comment]
Linux support in 2025 please

**Comment by JamesFieldist** [comment]
And windows please!

**Comment by YesThatGy** [comment]
Throwing my voice in the din as part of the choir seeking Windows and Linux support!

**Comment by houxd** [comment]
Asking for Linux support. Tnx!

**Comment by twixel-io** [comment]
Another voice for Windows & Linux... Please!

**Comment by o-glethorpe** [comment]
This is ridiculous. Could the FlutterFire team please address this? If you guys are not going to implement this, please just say so.

**Reply by NonymousMorlock** [reply]
Dear Google

**Comment by jmohanasundaram** [comment]
Asking for Linux support

**Comment by JohnGalt1717** [comment]
At this point I'd just settle for Messaging on windows....

**Reply by YesThatGy** [reply]
Yes, this please? 

If there is a reason why this cannot be implemented on Desktop platforms, can you please just let us know? Messaging apps is a key component to the function of our app and lack of this forces very, very awkward workarounds in order to support desktop clients. 

It's painful, it's expensive, and as far as I can tell, it's also un-necessary.

**Comment by stMerlHin** [comment]
Hello everyone. I notice that the team is not even looking into this issue to provide me with an answer. I wanted an app that would work across all popular platforms, but I figured I'd have to wait a few years for an answer. So I looked for an alternative and found [AWS Amplify](https://docs.amplify.aws/). They currently support all Flutter-compatible platforms through their Flutter plugins. With "AWS Amplify," I migrated my database from a non-relational database to a relational database, but the Flutter client code to interact with my backend hasn't changed. I'm now using a hybrid architecture. What I miss most about Firebase is its cache management. But guess what? I plan to use Firebase for mobile apps and Amplify for desktop and web to take advantage of Firebase's powerful caching system and, in the meantime, run my app on desktop and web by replicating data from Firebase to the Amplify backend via a cloud function or something similar. Amplify also has a caching system with "dataStore," but it's only available for "Amplify Gen 1" and isn't really compatible with desktop and web.

AWS Amplify has many features that Firebase doesn't and is compatible with all other services offered by AWS. It's also open source.

**Comment by o-glethorpe** [comment]
This made me choose Electron for my next project. It sucks, as expected, but at least I can deliver all the features my client paid me for. 
At this point, it seems to me that they dropped the ball when conceptualizing the "multi-platform" aspect of the Firebase Flutter SDK, in such a way that it can't be done unless a major rewrite happens first. Which looks like it will never happen, because their focus is AI now. The framework team itself seems to have been scaled back to a minimum—hence all the recent deprecations—so it's best to move on.
I once had a dream: to work fully with a Google stack—Angular, Firebase, Flutter, and Go where Firebase couldn't reach. Can you imagine if they had managed to make all of this work seamlessly together? But instead, what we got is a dysfunctional set of frameworks with poorly managed teams that don’t seem to talk to each other.

Now, for some projects, I have to ditch this amazing piece of tech that is Flutter and work with subpar tools like React Native, all because they messed it up.

**Comment by kito30** [comment]
Asking for Linux support

**Reply by richard457** [reply]
I do not understand why people choose Firebase, don't get me wrong, Firebase is awesome, but it has many downsides:

1. It is run by the corporation, which is a red flag. I will explain the last few months, MongoDB out of nowhere, and they closed their realm project
2. It is not open-source
3. Corporate never listen to small business
I am writing this here because it cost me time and money to wait for Firebase on Windows, and I could have gone to the market earlier if I had chosen something different, then I realized that choosing Firebase was wrong in the first place. reflecting to this thread where you won't get an answer clearly show me that choosing firebase is wrong totally wrong.

**Reply by JohnGalt1717** [reply]
Because if you want to do push notifications to Android you MUST use it. There is no other option. Which induces the use. Especially in Flutter which has few options for native notifications for other platforms other than Firebase.

**Reply by richard457** [reply]
Yes building custom solution is much safer than using firebase, for e.g I would use supabase and subscribe to a able that has notification then use service like https://pub.dev/packages/flutter_local_notifications to show notification.

**Reply by JohnGalt1717** [reply]
That doesn't solve offline push notifications where the app is closed.

Which is gated and on android REQUIRES the use of Firebase so you've already taken a massive dependency so it's easier to just use firebase for all offline push notifications for all platforms.

**Reply by richard457** [reply]
It depends, as of I knowledge offline doe not exist what you mean is that after an app woke up from offline firebase would still need internet to push notification to your app, same way if you come back online you would query notification you missed again.

**Reply by JohnGalt1717** [reply]
It doesn't depend. If you want notifications to be delivered while your app isn't running on Android, YOU MUST USE FIREBASE. There is no other way of doing so reliably.

Firebase integrates with Apple APNs and Web browsers as well.  It doesn't work with Windows or Linux and as a result, causes you to have to have at least 2 packages in your flutter app that do the same thing.

This isn't guessing. This isn't even a question. You MUST use firebase for notifications on Android. Full stop. The only question is if you want to subject yourself to the horrible 3rd party support on Flutter to get away from firebase on other platforms.

Anything else? Ya, sure, don't use firebase. There's lots of alternatives. But push notifications? You MUST.

**Reply by o-glethorpe** [reply]
> I do not understand why people choose Firebase, don't get me wrong, Firebase is awesome, but it has many downsides:
> 
> 1. It is run by the corporation, which is a red flag. I will explain the last few months, MongoDB out of nowhere, and they closed their realm project
> 2. It is not open-source
> 3. Corporate never listen to small business
>    I am writing this here because it cost me time and money to wait for Firebase on Windows, and I could have gone to the market earlier if I had chosen something different, then I realized that choosing Firebase was wrong in the first place. reflecting to this thread where you won't get an answer clearly show me that choosing firebase is wrong totally wrong.

Unlimited free projects (which is good for startups because you don't get a good revenue from the beginning).

It's cheaper (the cost model is simpler than Supabase because with Supabase you also need to pay to upgrade server instances; with Firebase, you don't need to worry about that if you don't use the new DataConnect).

It has all the services you commonly need when building an app, like push notifications, so you don't need another service just for that.

Of course, it all depends, but in my experience with both platforms, if you run the same project with a million users, you'll pay less with Firebase.

Do you disagree with any of this?

**Reply by richard457** [reply]
You forgot one major factor, the fact that it is not open-source and google can one day shut it down, also price wise I do not know if you attempted to do 1000 calls to firestore and see how price start to jump now imagine having millions of calls, on other things I will use firebase for non-major things such as notifications, crashlytics i.e the things I can switch easily but when it come to data a rather use supabase because I know if I want I would host entire data myself and have no vendor lock-in.

**Reply by JamesFieldist** [reply]
I could care less about Firestore. The only part of Firebase that is non-optional is Firebase messaging.

**Comment by richard457** [comment]
If you can pay the service for $20 I can build that for you :) hahaha

**Comment by channel101** [comment]
``` javascript

function input(the_team_is_working_on_this, there_is_no_progress)
if(the_team_is_working_on_this = false && there_is_no_progress = true) {
 alert("I didn't even expect it");
} else {
 alert("wow!, it can't not be!");
}

input(false, true);
```

result:

![clideo_editor_dbbea906ea594850b74f2ae9b383ff2b](https://github.com/user-attachments/assets/f8a84635-798c-4824-951c-6c0e05f4aaa8)

**Comment by o-glethorpe** [comment]
We’ll never have full Firebase support across all environments.

**Comment by marcocatelli** [comment]
can't understand why a company like google, which makes heavy use of linux, is not supporting linux on their flagship developer service (firebase) in their flagship developer environment (flutter). That's crazy

**Comment by jamespet77** [comment]
OMG...  what year is it and Google is not supporting Linux with this yet?  Seriously.  this is an embarrassment to flutterfire developers.
